### PR TITLE
Inside LTO improvements.

### PR DESCRIPTION
**Pull request by ehuss**
This is just a random post about some recent performance and disk-space improvements that have recently been introduced. I kinda felt someone might find it interesting.

### CODE REVIEW

**Comment by ehuss** [comment]
@alexcrichton or @nnethercote, if either of you could look this over for any major errors, I would appreciate it!

File: posts/inside-rust/2020-06-24-lto-improvements.md [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+Thanks to the work of [Nicholas Nethercote] and [Alex Crichton], there have been some recent improvements that reduce the size of compiled libraries, and improves the compile-time performance, particularly when using LTO. This post dives into some of the details of what changed, and an estimation of the benefits.
+
+These changes have been added incrementally over the past three months, with the latest changes landing just a few days ago on the nightly channel. The bulk of the improvements will be found in the 1.46 stable release. It would be great for any projects that use LTO to test it out on the nightly channel and report any issues that arise.
+
+[Nicholas Nethercote]: https://github.com/nnethercote
+[Alex Crichton]: https://github.com/alexcrichton/
+
+## Background
+
+When compiling a library, `rustc` saves the output in an `rlib` file which is an [archive file]. This has historically contained these two things (among others):
```
Comment (nnethercote) [inline posts/inside-rust/2020-06-24-lto-improvements.md:18]: Not two but three things: there is metadata as well.

File: posts/inside-rust/2020-06-24-lto-improvements.md [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+If the project is using LTO, then Cargo will instruct `rustc` to not place object code in the rlib files, avoiding the expensive code generation step. This should improve the build time when building from scratch, and reduce the amount of disk space used.
+
+Two `rustc` flags are now available to control how the rlib is constructed:
+
+* [`-C linker-plugin-lto`] causes `rustc` to only place bitcode in the `.o` files, and skips code generation. Cargo uses this when the rlib is only intended for use with LTO. This can also be used when doing cross-language LTO.
+* [`-C embed-bitcode=no`] causes `rustc` to avoid placing bitcode in the rlib altogether. Cargo uses this when LTO is not being used, which reduces some disk space usage.
+
+Additionally, the method in which bitcode is embedded in the rlib has changed. Previously, `rustc` would place compressed bitcode as a `.bc.z` file in the rlib archive. Now, the bitcode is placed as an uncompressed section within each `.o` [object file] in the rlib archive. This avoids a small performance hit for compressing the bitcode, and also matches the standard format used by clang.
```
Comment (nnethercote) [inline posts/inside-rust/2020-06-24-lto-improvements.md:40]: The performance effect here is a mixed bag. Sometimes it's a speedup because we avoid the compression cost. Sometimes it's a slowdown because more data must be written. The combined effect is hard to predict and depends on the program. Avoiding the compression certainly makes the compiler simpler, though.

File: posts/inside-rust/2020-06-24-lto-improvements.md (hunk 1 above)
Comment (nnethercote) [reply posts/inside-rust/2020-06-24-lto-improvements.md:18]: Maybe the metadata is covered by the "(among others)", but given that it's the only thing in "(among others)", it's probably just worth mentioning outright.

File: posts/inside-rust/2020-06-24-lto-improvements.md (hunk 2 above)
Comment (alexcrichton) [reply posts/inside-rust/2020-06-24-lto-improvements.md:40]: It's also probably worth pointing out that "standard format" is a bit of a stretch as we discovered because only a few linkers actually work with the default output, we had to frob with some global inline asm to get the sections to work out on all platforms.

In reality this "standard format" was only created for iOS where they ship bitcode by default but wanted a mode of shipping both for debugging and such. We're kinda just piggy-backing on it at this point. In any case the words written down here are correct, so I'll leave it up to you if you'd like to add more context!

File: posts/inside-rust/2020-06-24-lto-improvements.md [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+## Improvements
+
+The following is a summary of improvements observed on a small number of real-world projects of small and medium size. Improvements of a project will depend heavily on the code, optimization settings, operating system, environment, and hardware. These were recorded with the 2020-06-21 nightly release on Linux with parallel job settings between 2 and 32.
+
+The performance wins for debug builds were anywhere from 0% to 4.7% faster. Larger binary crates tended to fair better than smaller library crates.
+
+LTO builds were recorded anywhere from 4% to 20% faster. Thin LTO faired consistently better than fat LTO.
+
+The number of parallel jobs also had a large impact on the amount of improvement. Lower parallel job counts saw substantially more benefit than higher ones. A project built with `-j2` can be 20% faster, whereas the same project at `-j32` would only be 1% faster. Presumably this is because the code-generation phase benefits from higher concurrency, so it was taking a relatively smaller total percentage of time.
```
Comment (alexcrichton) [inline posts/inside-rust/2020-06-24-lto-improvements.md:50]: "Lower parallel job counts saw substantially more benefit than higher ones" -- oh man that's pipelining in action I think!

File: posts/inside-rust/2020-06-24-lto-improvements.md [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+When compiling a library, `rustc` saves the output in an `rlib` file which is an [archive file]. This has historically contained these two things (among others):
+
+* Object code, which is the result of code generation. This is used during regular linking.
+* [LLVM bitcode], which is a binary representation of LLVM's intermediate representation. This can be used for [Link Time Optimization] (LTO).
+
+LTO is an optimization technique that can perform whole-program analysis. It analyzes all of the bitcode from every library at once, and performs optimizations and code generation. `rustc` supports several forms of LTO:
+
+* Fat LTO. This performs "full" LTO, which can take a long time to complete and may require a significant amount of memory.
+* [Thin LTO]. This is a lightweight version of "fat" LTO that can achieve similar performance improvements while taking much less time to complete.
```
[the remark this answers is not in the document]
Comment (panaman67) [reply posts/inside-rust/2020-06-24-lto-improvements.md:26]: @alexcrichton  Because of the phrasing of this sentence, is there ever a time where fat LTO is better than thin LTO?

[the remark this answers is not in the document]
Comment (alexcrichton) [reply posts/inside-rust/2020-06-24-lto-improvements.md:26]: Heh I wondered the same thing when ThinLTO came along! Fat LTO, however, produces smaller binaries because it can extremely aggressively internalize functions. It can also produce slightly faster binaries since total knowledge in a compiler is typically better than fragmented knowledge. ThinLTO's primary purpose is figure out what to inline across codegen units, while Fat LTO basically says "eh just do your normal thing but here's all the code in the world"

[the remark this answers is not in the document]
Comment (panaman67) [reply posts/inside-rust/2020-06-24-lto-improvements.md:26]: That makes sense. Is there a way we can rephrase this to make this point more clear? The way it stands now makes it seem like thin is always better. "Similar performance while taking much less time"

File: posts/inside-rust/2020-06-24-lto-improvements.md (hunk 1 above)
Comment (ehuss) [reply posts/inside-rust/2020-06-24-lto-improvements.md]: Yea, it seemed awkward when I wrote it.  I was attempting to focus on just the things that changed. I went ahead and just added the metadata as the 3rd major part.

File: posts/inside-rust/2020-06-24-lto-improvements.md (hunk 2 above)
Comment (ehuss) [reply posts/inside-rust/2020-06-24-lto-improvements.md]: Thanks for the info!  Updated with a little extra information.

File: posts/inside-rust/2020-06-24-lto-improvements.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (ehuss) [reply posts/inside-rust/2020-06-24-lto-improvements.md]: I've tweaked the wording a little to try to make it clearer.

I think as with any optimization settings, every project will need to experiment with different settings to see what works best for them.

File: posts/inside-rust/2020-06-24-lto-improvements.md [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+layout: post
+title: "Disk space and LTO improvements"
+author: Eric Huss
+description: "Disk space and LTO improvements"
+team: the Cargo team <https://www.rust-lang.org/governance/teams/dev-tools#cargo>
+---
+
+Thanks to the work of [Nicholas Nethercote] and [Alex Crichton], there have been some recent improvements that reduce the size of compiled libraries, and improves the compile-time performance, particularly when using LTO. This post dives into some of the details of what changed, and an estimation of the benefits.
+
+These changes have been added incrementally over the past three months, with the latest changes landing just a few days ago on the nightly channel. The bulk of the improvements will be found in the 1.46 stable release. It would be great for any projects that use LTO to test it out on the nightly channel and report any issues that arise.
```
[the remark this answers is not in the document]
Comment (ehuss) [reply posts/inside-rust/2020-06-24-lto-improvements.md:11]: Sounds good, added the date for the most recent fixes.

**Comment by nikomatsakis** [comment]
@ehuss let me know when you want to post!

**Comment by ehuss** [comment]
This should be ready to go!

**Comment by Mark-Simulacrum** [comment]
Want to bump the date in the file to Monday and we can merge tomorrow?
